snf-cyclades-app/synnefo/api/keypairs.py

```diff
@@ -0,0 +1,160 @@
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import json
+
+from django.conf import settings
+from django.conf.urls import patterns
+from django.http import HttpResponse, HttpResponseServerError
+
+from snf_django.lib.astakos import get_user
+from snf_django.lib import api
+from snf_django.lib.api import utils, faults
+
+from synnefo.userdata.models import PublicKeyPair
+from synnefo.userdata.util import generate_keypair, SUPPORT_GENERATE_KEYS
+from synnefo.api import util
+
+from logging import getLogger
+
+log = getLogger(__name__)
+
+urlpatterns = patterns(
+    'synnefo.api.keypairs',
+    (r'^(?:/|.json|.xml)?$', 'demux'),
+    (r'^/([\w-]+)(?:/|.json|.xml)?$', 'keypair_demux'),
+)
+
+
+def keypair_to_dict(keypair, details=False):
+    d = {
+        'name': keypair.name,
+        'fingerprint': keypair.fingerprint,
+        'public_key': keypair.content
+    }
+    if details is True:
+        d['deleted'] = keypair.deleted
+        d['id'] = keypair.id
+        d['created_at'] = keypair.created_at.isoformat()
+        d['deleted_at'] = None
+        if keypair.deleted_at is not None:
+            d['deleted_at'] = keypair.deleted_at.isoformat()
+        d['updated_at'] = None
+        if keypair.updated_at is not None:
+            d['updated_at'] = keypair.updated_at.isoformat()
+    return {'keypair': d}
+
+
+def demux(request):
+    if request.method == 'GET':
+        return list_keypairs(request)
+    elif request.method == 'POST':
+        return create_new_keypair(request)
+    else:
+        return api.api_method_not_allowed(request,
+                                          allowed_methods=['GET', 'POST'])
+
+
+def keypair_demux(request, keypair_name):
+    if request.method == 'GET':
+        return get_keypair(request, keypair_name)
+    if request.method == 'DELETE':
+        return delete_keypair(request, keypair_name)
+    else:
+        return api.api_method_not_allowed(request,
+                                          allowed_methods=['GET', 'DELETE'])
+
+
+@api.api_method(http_method='GET', user_required=True, logger=log)
+def list_keypairs(request):
+    active_keypairs = PublicKeyPair.objects.filter(user=request.user_uniq).all()
+    keypairs = [keypair_to_dict(keypair) for keypair in
+                active_keypairs.order_by('name')]
+    data = json.dumps({'keypairs': keypairs})
+    return HttpResponse(data, status=200)
+
+
+@api.api_method(http_method='POST', user_required=True, logger=log)
+def create_new_keypair(request):
```

Comment:
Every API view that updates the db should run in a transaction. You have to decorate them with `@transaction.commit_on_success` (`from synnefo.db import transaction`).